mrbgems/mruby-catch/mrblib/catch.rb

```diff
@@ -9,14 +9,14 @@ def initialize(tag, val)
 
 module Kernel
   def catch(tag=Object.new, &block)
-    block.call(tag)
-  rescue ThrowCatchJump => e
-    unless e._tag.equal?(tag)
-      raise e
-    end
-    return e._val
+    # A double closure is required to make the nested `catch` distinguishable
+    # and because `break` goes back to `proc->upper`.
+    -> { -> { block.call(tag) }.call }.call
   end
   def throw(tag, val=nil)
-    raise ThrowCatchJump.new(tag, val)
+    __throw(tag, val)
+    raise UncaughtThrowError.new(tag, val)
```

Comment:
`UncaughtThrowError#initialize` is defined in `catch.rb`, so it was separated to avoid reentry to the VM when raising an exception.

Reply:
This reentrance should not cause a new problem because every exception class generation from `mrb_raise()` and `mrb_raisef()` calls `mrb_obj_new()` that calls VM recursively. Avoiding one call here does not improve overall situation.

Reply:
Thank you for your reply. I understood that.